File: Combinations - Leetcode 77.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        ans, sol = [], []

        def backtrack(x):
            if len(sol) == k:
                ans.append(sol[:])
                return

            left = x
            still_need = k - len(sol)

            if left > still_need:
                backtrack(x - 1)

            sol.append(x)
            backtrack(x - 1)
            sol.pop()

        backtrack(n)
        return ans

# Time Complexity: O(n choose k)
# Space Complexity: O(n)

File: Combinations - Leetcode 77/Combinations - Leetcode 77.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
    const ans = [];
    const sol = [];
    backtrack(n, k, sol, ans);
    return ans;
};

function backtrack(x, k, sol, ans) {
    if (sol.length === k) {
        ans.push([...sol]);
        return;
    }

    if (x > k - sol.length) {
        backtrack(x - 1, k, sol, ans);
    }

    sol.push(x);
    backtrack(x - 1, k, sol, ans);
    sol.pop();
}
